DOCS-10142 updated ping federate policy (#336)
mulesoft/docs-api-manager: modules/ROOT/pages/policy-ping-federate.adoc

>s| Policy Name | PingFederate OAuth 2.0 Token Enforcement
>s|Summary      | Allows access only to authorized client applications
>s|Category | Security
>s| First Mule version available | v4.1.1
.4+>.^s| Returned Status Codes
|400 - Invalid token
|401 - Unauthorized access or error when connecting to the authorization server
|403 - Forbidden, invalid client application credentials
|500 - Bad response from authorization server, or WSDL SOAP fault error.
|===

> The PingFederate OAuth 2.0 Token Enforcement policy restricts access to a protected resource to only those HTTP requests that provide a valid Oauth2 token belonging to a client application with API access. The policy does not generate tokens but only validates them. 

Reviewer: ```suggestion
The PingFederate OAuth 2.0 Token Enforcement policy restricts access to a protected resource to only those HTTP requests that provide a valid OAuth2 token belonging to a client application with API access. The policy does not generate tokens but only validates them. 
```

Author: Done! Thanks!

=== Token Caching

After the system validates a token, that token is cached (by default) until it expires, to improve the performance of the policy. In certain cases, such as in the case of revoked tokens, you might want to minimize or even disable caching.

To control the time taken to cache the tokens before retrying a validation against the token validation endpoint, specify the following property when starting the Mule runtime engine:

`anypoint.platform.max_federation_expiration_time=<expiration time in seconds; specify 0 to disable caching>`

Alternatively, you can specify this parameter in the `wrapper.conf` file.
 
To control the number of tokens that can be cached simultaneously, specify the following property when starting the Mule runtime engine:
 
`anypoint.platform.ping_federate_cache_max_size=<amount of tokens to be cached; default value is 10.000>`

Alternatively, you can specify this parameter in the `wrapper.conf` file.

== Configuring the Expose Headers Option

If you configured the `Expose Headers` option in the policy and if the application (for example, out-of-the-box proxies) uses the HTTP requester, the properties are redirected to the backend as headers. 

> For each field, the original request to the protected resource is enriched with the HTTP headers using the format: X-AGW- + <key>=<value>. For the example response from the previous section, the following headers are added:

Reviewer: ```suggestion
For each field, the original request to the protected resource is enriched with the HTTP headers using the format X-AGW- + <key>=<value>. For the example response from the previous section, the following headers are added:
```

Author: Done! Thanks!

[source,text,linenums]

> ---

Reviewer: Should these be formatted as syntax (` `)?

Author: Don't see this in my file. I think it was removed.

X-AGW-uid=john.doe
X-AGW-mail=john.doe@example.com
X-AGW-token_type=Bearer
X-AGW-expires_in=3600
---

For further processing within the flow, the original unparsed response from the validation endpoint is made available. The information is stored in an authentication object. 

For example, if a `mail` field is returned to the policy, you can access the value of the field by using the following DataWeave 2.0 expression:

`#[authentication.properties.userProperties.mail]`

You can obtain the client ID of the OAuth2 token by using the following DataWeave 2.0 expression:

`#[authentication.principal]`

== Communicating with the Token Validation Endpoint Using a Proxy

> You can enable the PingFederate OAuth 2.0 Token Enforcement policy to use the gateway proxy settings by specify the following property when starting the Mule runtime engine: 

Reviewer: ```suggestion
You can enable the PingFederate OAuth 2.0 Token Enforcement policy to use the gateway proxy settings by specifying the following property when starting the Mule runtime engine: 
```

Author: Done! Thanks!

`anypoint.platform.external_authentication_provider_enable_proxy_settings=<true|false(default)>`

When you enable this property, the policy uses the Mule proxy settings, if you have specified the following parameters:

* `anypoint.platform.proxy_host=localhost`
* `anypoint.platform.proxy_port=8080`

=== Configuring Parameters for Mule Applications

When you apply the policy to your API from the UI, a list of parameters is displayed:

[%header%autowidth.spread,cols="30,40,30"]
|===
| Element | Description | Example
| Scopes | Displays a space-separated list of supported scopes. | `READ`, `WRITE`, `READ and WRITE`

> | Scope Validation Criteria | Determines if the token must contain all defined scopes or just one.

Reviewer: ```suggestion
| Scope Validation Criteria | Determines whether the token must contain all defined scopes or just one.
```

Author: Done! Thanks!

Values: `Contains all scopes` and `Contains any scope` | 
If the ‘Contains all scopes’ value is selected, the tokens with the scopes `READ` and `WRITE` are accepted. If the ‘Contains any scopes’ value is selected, the tokens with the scope `READ`, `WRITE`, and `READ AND WRITE` are accepted. 
| Validate TLS Certificate | Enables third-party authentication server TLS validation. | If selected, the communication with the token validation endpoint uses the TLS protocol.
| Expose Headers | In a proxy scenario, specifies whether headers must be exposed in the request to the backend. User properties returned by the federation server are prefixed with 'X-AGW-' and sent as headers to the backend. | See the <<leveraging-TVE-information,Leveraging Token Validation Endpoint Information for Mule Applications>> for more information about this element.

> | Skip Client Id Validation | Skips the client application's API contract validation. | If selected, the Step 4 of the diagram in the <<how-this-policy-works,How This Policy Works>> section is skipped.

Reviewer: ```suggestion
| Skip Client Id Validation | Skips the client application's API contract validation. | If selected, Step 4 of the diagram in the <<how-this-policy-works,How This Policy Works>> section is skipped.
```
Not a good idea to refer to the content of one topic in another, because this creates intertopic dependencies.

Author: Done! Thanks!

|===

== FAQs

> *The PingFederate OAuth 2.0 Token Enforcement does not appear in the list of policies to apply to the API.*

Reviewer: Parallelism/make a question:
```suggestion
*Why doesn't the PingFederate OAuth 2.0 Token Enforcement appear in the list of policies to apply to the API?*
```

Author: Done! Thanks!

If the policy does not appear in the policies list, ensure that you have configured the PingFederate client provider in Access Management. For information, see
xref:access-management::conf-client-mgmt-pf-task.adoc[Configure PingFederate Client Management]. If you are using xref:configure-multiple-credential-providers.adoc[multiple IdPs], verify that the PingFederate client is properly configured in the API.

*Can I generate OAuth2 tokens with the policy?*

No. The policy only validates the tokens.

*How does the policy manage the credentials to communicate with the validation endpoint?*

This information is part of the policy configuration and is stored on a disk. If you want this information to be stored with encryption, xref:api-gateway-encryption-mule4.adoc[configure gateway encryption in your runtime]. 

> *How are the OAuth tokens cached?*

Reviewer: Where
```suggestion
*Where are the OAuth tokens cached?*
```

Author: Done! Thanks!

OAuth2 tokens are cached only in memory and are never written to disk.

*Does the policy communicate with the Anypoint Platform on a per-request basis?*

No, the policy uses client applications previously retrieved by the Mule runtime engine.

> This enables the policy to continue working even if the connection with the Management plane is lost. If you want the client application information that is written on disk to be encrypted, xref:api-gateway-encryption-mule4.adoc[configure gateway encryption in your runtime].

Reviewer: "control plane," per Term Presentation List:
```suggestion
This enables the policy to continue working even if the connection with the canagement plane is lost. If you want the client application information that is written on disk to be encrypted, xref:api-gateway-encryption-mule4.adoc[configure gateway encryption in your runtime].
```

Author: Done! Thanks!
